zerver/migrations/0443_backfill_subscription_auditlogs.py

```diff
@@ -0,0 +1,117 @@
+        last_id = Message.objects.aggregate(Max("id"))["id__max"]
+        if last_id is None:
+            # During initial realm creation, there might be 0 messages in
+            # the database; in that case, the `aggregate` query returns
+            # None.  Since we want an int for "beginning of time", use -1.
+            last_id = -1
+        return last_id
+
+    for stream in Stream.objects.all():
+        subscribed_user_ids = set(
+            Subscription.objects.filter(recipient_id=stream.recipient_id).values_list(
+                "user_profile_id", flat=True
+            )
+        )
+        user_ids_in_audit_log = set(
+            RealmAuditLog.objects.filter(
+                realm=stream.realm,
+                event_type__in=[
+                    301,  # RealmAuditLog.SUBSCRIPTION_CREATED
+                    302,  # RealmAuditLog.SUBSCRIPTION_ACTIVATED
+                    303,  # RealmAuditLog.SUBSCRIPTION_DEACTIVATED
+                ],
+                modified_stream=stream,
+            )
+            .distinct("modified_user_id")
+            .values_list("modified_user_id", flat=True)
+        )
+
+        user_ids_missing_events = subscribed_user_ids - user_ids_in_audit_log
+        if not user_ids_missing_events:
+            continue
+
+        backfills = []
+        for user_id in sorted(user_ids_missing_events):
+            print(
+                f"Backfilling subscription event for {user_id} in {stream.name} in {stream.realm.string_id}"
+            )
+            aggregated = UserMessage.objects.filter(
+                user_profile_id=user_id,
+                message__recipient=stream.recipient_id,
+            ).aggregate(
+                earliest_date=Min("message__date_sent"),
+                earliest_message_id=Min("message_id"),
+            )
+
+            # Assume we subscribed right before the first message we
+            # saw -- or, if we don't see any, right now.  This makes
+            # this safe for streams which do not have shared history.
+            if aggregated["earliest_message_id"] is not None:
+                event_last_message_id = aggregated["earliest_message_id"] - 1
+            else:
+                event_last_message_id = get_last_message_id()
+            if aggregated["earliest_date"] is not None:
+                event_time = aggregated["earliest_date"]
+            else:
+                event_time = timezone_now()
+
+            log_event = RealmAuditLog(
+                event_time=event_time,
+                event_last_message_id=event_last_message_id,
+                backfilled=True,
+                event_type=301,  # RealmAuditLog.SUBSCRIPTION_CREATED
+                realm_id=stream.realm_id,
+                modified_user_id=user_id,
+                modified_stream_id=stream.id,
+            )
+            backfills.append(log_event)
+        RealmAuditLog.objects.bulk_create(backfills)
```

Comment:
I suppose we don't know whether this is possible with the bug in question, but if the user's subscription is currently deactivated, then I think we might need to create a `SUBSCRIPTION_DEACTIVATED` entry too. If we think that's not possible, I suppose we could crash should that occur? In theory we could figure out what entry to put based on the highest `UserMessage` ID matching, but I'm not sure it's worth writing/testing that code if we don't think it can happen?

Reply:
I don't see how they could need a `SUBSCRIPTION_DEACTIVATED` if they currently have a subscription row?